### PR TITLE
Add relation.afterAttach, relation.afterDetach events in BelongsToMany relations

### PR DESCRIPTION
**Pull request by kharanenka**
Related [issue](https://github.com/octobercms/october/issues/3675).
I add new events for BelongsToMany relation
  * **relation.afterAttach**
  * **relation.afterDetach**
Usage example:
```php
Product::extend(function ($obModel) {
    $obModel->bindEvent('relation.afterAttach', function ($sRelationName, $arElementIDList, $arInsertData) {
        //$sRelationName == 'category'
        //$arElementIDList == [1,3,15]
        //$arInsertData = [['product_id' => 10, 'category_id' => 1],['product_id' => 10, 'category_id' => 3],['product_id' => 10, 'category_id' => 15, 'pivot_field' => 'test']];
    });
    $obModel->bindEvent('relation.afterDetach', function ($sRelationName, $arElementIDList) {
        //$sRelationName == 'category'
        //$arElementIDList == [1,3,15]
    });
});
```

### CODE REVIEW

**Comment by LukeTowers** [comment]
Model events must be prefixed with `model.` (so `model.relation.afterAttach`). Before events are needed here as well with the ability for return false in their listeners to prevent attachment, and inline event documentation for the API docs needs to be added. Also, October does not use the variable naming of (oneLetterType)$var, so please don't name things `$sVar` or `$arArray`)

**Comment by kharanenka** [comment]
Do I need to change $insertData array in "before" event?

**Comment by kharanenka** [comment]
Where can i find example of the API docs for events?

File: src/Database/Relations/BelongsToMany.php [hunk 1]
```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
...
+        $insertData = $this->formatAttachRecords($this->parseIds($id), $attributes);
+        $elementIDList = array_pluck($insertData, $this->relatedPivotKey);
+
+        /**
+         * @event model.relation.beforeAttach
+         * Called before creating a new relation between models (only for BelongsToMany relation)
+         *
+         * Example usage:
+         *
+         *     $model->bindEvent('model.relation.beforeAttach', function ($relationName, $elementIDList, $insertData) use (\October\Rain\Database\Model $model) {
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: Please use type hinting in the examples to let users know what type of argument they're getting, i.e. `(string) $relation, (array) $ids, (array) $data`

File: src/Database/Relations/BelongsToMany.php [hunk 2]
```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
...
     }
 
+    /**
+     * @param mixed $id
+     * @param array $attributes
+     * @param bool  $touch
+     */
+    public function attach($id, array $attributes = [], $touch = true)
+    {
+        $insertData = $this->formatAttachRecords($this->parseIds($id), $attributes);
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: This is fine as just `$data`

File: src/Database/Relations/BelongsToMany.php [hunk 3]
```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
...
+        $elementIDList = array_pluck($insertData, $this->relatedPivotKey);
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: Could you give me more information on what exactly this variable holds? `$elementIDList` isn't very clear to me what's going on.

File: src/Database/Relations/BelongsToMany.php (hunk 2 above)
Comment (kharanenka) [reply src/Database/Relations/BelongsToMany.php]: I do not agree, "$data" variable name is too abstract.

File: src/Database/Relations/BelongsToMany.php (hunk 3 above)
Comment (kharanenka) [reply src/Database/Relations/BelongsToMany.php]: I receive IDs of elements that have been attached.
For example:
```
$obProduct->attach(1); //$elementIDList == [1];
$obProduct->attach([1,3]); //$elementIDList == [1,3];
$obProduct->attach([1 => ['pivot_field' => 'test'], 2, 5, 10]); //$elementIDList == [1, 2, 5,10];
```

File: src/Database/Relations/BelongsToMany.php [hunk 4]
```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
...
+         *     $model->bindEvent('model.relation.beforeAttach', function (string $relationName, array $elementIDList, array $insertData) use (\October\Rain\Database\Model $model) {
+         *         if (!$model->isRelationValid($elementIDList)) {
+         *             throw new \Exception("Invalid relation!");
+         *             return false;
+         *         }
+         *     });
+         *
+         */
+        $eventDataList = $this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $elementIDList, $insertData]);
+        if (!empty($eventDataList)) {
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: No need for this $eventDataList handling code, the `$halt` parameter should be passed to fireEvent instead simplifying to 
```php
if ($this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $elementIDList, $insertData], true) {
    return;
}
```

See http://octobercms.com/docs/api/october/rain/support/traits/emitter#fireevent

File: src/Database/Relations/BelongsToMany.php [hunk 5]
```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
         return $model;
     }
 
+    /**
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: Please provide method summaries for both methods added in their docblocks

File: src/Database/Relations/BelongsToMany.php (hunk 3 above)
Comment (LukeTowers) [reply src/Database/Relations/BelongsToMany.php]: Please change to `$attachedIDs` instead then

File: src/Database/Relations/BelongsToMany.php (hunk 5 above)
Comment (kharanenka) [reply src/Database/Relations/BelongsToMany.php]: Sorry, I don't understand you.

Comment (LukeTowers) [reply src/Database/Relations/BelongsToMany.php]:
```php
/** 
  * Summary of what method does goes here
  * 
  * @param type $var blah blah blah
  * @return void
  */
```

File: src/Database/Relations/BelongsToMany.php (hunk 3 above)
Comment (LukeTowers) [reply src/Database/Relations/BelongsToMany.php]: @kharanenka not $attachedIDList, $attachedIDs. October does not include the type of variable in the name of the variable anywhere in it's code base so I don't want to start doing it here.

File: src/Database/Relations/BelongsToMany.php [hunk 6]
```diff
@@ -100,6 +100,100 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
...
+         *
+         *     $model->bindEvent('model.relation.beforeAttach', function (string $relationName, array $attachedIDList, array $insertData) use (\October\Rain\Database\Model $model) {
+         *         if (!$model->isRelationValid($attachedIDList)) {
+         *             throw new \Exception("Invalid relation!");
+         *             return false;
+         *         }
+         *     });
+         *
+         */
+        if ($this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $attachedIDList, $insertData], true)) {
```
Comment (LukeTowers) [inline src/Database/Relations/BelongsToMany.php]: See https://github.com/octobercms/library/blob/master/src/Database/Model.php#L728 for how to implement halting events. You still need to check to see if it has returned false

File: src/Database/Relations/BelongsToMany.php (hunk 5 above)
Comment (kharanenka) [reply src/Database/Relations/BelongsToMany.php]: fixed

**Comment by daftspunk** [comment]
LGTM; Thanks @kharanenka and @LukeTowers

**Comment by kharanenka** [comment]
Hi @daftspunk! When can we get this issue in kernel? I need to plan the release of plugin updates  with using this improvement.

**Comment by kharanenka** [comment]
I noticed the following:
When we call detach() method without parameters, the event will not send a list of the IDs of the related items. It's not critical, but I'm not sure if it's right.

**Comment by LukeTowers** [comment]
@kharanenka feel free to submit a PR to fix it

**Comment by kharanenka** [comment]
I create [PR](https://github.com/octobercms/library/pull/349) with fix.
I added the receipt of a list of related elements, if we call detach() method without parameters

**Comment by chrisvidal** [comment]
what about the event when an existing pivot relationship is updated?

**Comment by kharanenka** [comment]
Hi @chrisvidal! 
```php
if ($this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $attachedIDList, $insertData], true) === false) {
            return;
        }
```
$insertData is pivot data.

**Comment by chrisvidal** [comment]
Hi
$insertData  is always null in my cases

**Pull request merged**